rust_src/src/xdisp.rs

```diff
@@ -37,4 +41,80 @@ pub fn trace_redisplay(arg: Option<InteractiveNumericPrefix>) {
     }
 }
 
+fn invisible_prop(propval: LispObject, list: LispCons) -> EmacsInt {
+    let find_prop = |propelt: LispObject| {
+        for tem in list.iter_cars(LispConsEndChecks::off, LispConsCircularChecks::off) {
+            if propelt.eq(tem) {
+                return Some(1);
+            }
+            if tem.is_cons() {
+                let (car, cdr) = tem.into();
+                if propelt.eq(car) {
+                    return Some(if cdr.is_nil() { 1 } else { 2 });
+                }
+            }
+        }
+        None
+    };
+
+    if let Some(val) = find_prop(propval) {
+        return val;
+    }
+
+    if propval.is_cons() {
+        for propelt in propval.iter_cars(LispConsEndChecks::off, LispConsCircularChecks::off) {
+            if let Some(val) = find_prop(propelt) {
+                return val;
+            }
+        }
+    }
+
+    0
+}
+
+// originally the macro TEXT_PROP_MEANS_INVISIBLE from intervals.h
+fn text_prop_means_invisible(prop: LispObject) -> EmacsInt {
```

Comment:
Is there a reason to keep the C version and this version?

Reply:
Yes, `TEXT_PROP_MEANS_INVISIBLE `is used in a bunch of other places, in xdisp.c and keyboard.c. Presumably this could be moved into intervals.rs, but I'm not sure if Rust is okay with that sort of circular dependency (since xdisp.rs will need to load it from intervals.rs, but the function calls `invisible_prop` from xdisp.rs)?

Reply:
What you could do is write a small Rust function which exposes this function to C using the expected C name. This way when there are no more callers the C version can be removed and only the Rust version remains.

Yes, Rust handles that kind of circular depends just fine. I wish Python did!

Reply:
Is it possible to have the Rust function called in place of the C macro, or will I need to change the C macro to just call a function that's declared `extern` & implemented in Rust?

Reply:
Macros are resolved long before linking the program, so to change what a macro does you will need to change the definition of the macro :)

Reply:
👍 cool, that's what I figured, but I wasn't sure if there was some sort of Deep Rust Magic that would let it replace a macro :grin:

Reply:
@db48x true, but in this case, does it _need_ to be a macro?

We can keep two copies of the logic, no worries there. It would be my preference to avoid that when it makes sense.

Reply:
Dumb question -- I've added an `intervals.rs` file in `rust_src/src`, but when I try to add `use crate::intervals::text_prop_means_invisible` to xdisp.rs, it can't find the crate. What do I need to do to add the new file to the project?

Reply:
ah, okay; figured out I need to add it to src/lib.rs